dataset-service/services/observation_service.go

```
@@ -0,0 +1,398 @@
+}
+
+// NewObservationService instantiates ObservationService
+func NewObservationService(
+	services *Services,
+	deploymentConfig common_config.DeploymentConfig,
+	observationServiceConfig config.ObservationServiceConfig,
+) ObservationService {
+	return &observationService{
+		services:                 services,
+		gcpProject:               observationServiceConfig.GCPProject,
+		deploymentConfig:         deploymentConfig,
+		observationServiceConfig: observationServiceConfig,
+	}
+}
+
+func (o *observationService) CreateService(
+	caramlProjectName string,
+	config *timberv1.ObservationServiceConfig,
+) (*timberv1.ObservationServiceResponse, error) {
+	// Read chart
+	chart, err := readChart()
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if chart is installable
+	validInstallableChart, err := isChartInstallable(chart)
+	if !validInstallableChart {
+		log.Info(err)
+	}
+
+	// Retrieve computed chart values based on default values and request body values
+	updatedChartValues, err := getChartValues(config, o.gcpProject, caramlProjectName, o.deploymentConfig, o.observationServiceConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate configuration required to run helm installation, this is dependent on storage backend.
+	settings := cli.New()
+	actionConfig := new(action.Configuration)
+	err = actionConfig.Init(settings.RESTClientGetter(), caramlProjectName, HELM_DRIVER, log.Infof)
+	if err != nil {
+		return nil, err
+	}
+
+	// Postfix helm release name with provided Service name as a CaraML project could have multiple different services
+	projectReleaseName := fmt.Sprintf("%s-%s", RELEASE_NAME, config.GetServiceName())
+
+	// Initialize helm installation
+	installation := action.NewInstall(actionConfig)
+	installation.ReleaseName = projectReleaseName
+	installation.Namespace = caramlProjectName
+	installation.CreateNamespace = true
+
+	// Trigger helm installation
+	release, err := installation.Run(chart, updatedChartValues)
```

does it wait until success?

---

It doesn't, I think we can use a goroutine to track this status and write the status to DB whenever we integrate the persistence layer to Dataset Service, wdyt?

---

yes that sounds good, side question if there are 2 requests with the same payload for create observation service, the second request would be fail?

---

I think we can track the installation state changes in the DB and block an attempt to deploy an already "pending deployment" using that. Wdyt?

---

we can do that once integrate the db. I just want to understand for current state (no db yet) the same payload to create observation service, the second request would be fail, since the chart already installed, is that correct?

---

>if there are 2 requests with the same payload for create observation service, the second request would be fail?

Yes, it should fail. Ideally, this function should first check the DB if such a helm installation exists and if it does, it should throw an error.